### ISSUE TITLE
LanguageServerProjectComponent.setupServer uses disposed component

### ISSUE THREAD

**Issue by ivanporty** [GoogleCloudPlatform/cloud-code-intellij#2694]
```
com.intellij.util.IncorrectOperationException: Sorry but parent: Project (name=1.14 ***)  is being disposed so the child: false will never be disposed
	at com.intellij.openapi.util.ObjectTree.register(ObjectTree.java:65)
	at com.intellij.openapi.util.Disposer.register(Disposer.java:78)
	at com.intellij.util.containers.DisposableWrapperList.createDisposableWrapper(DisposableWrapperList.java:255)
	at com.intellij.util.containers.DisposableWrapperList.add(DisposableWrapperList.java:62)
	at com.intellij.util.EventDispatcher.addListener(EventDispatcher.java:155)
	at com.intellij.openapi.editor.impl.event.EditorEventMulticasterImpl.addDocumentListener(EditorEventMulticasterImpl.java:120)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$setupServer$2.invokeSuspend(LanguageServerProjectComponent.kt:109)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$setupServer$2.invoke(LanguageServerProjectComponent.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:156)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent.setupServer(LanguageServerProjectComponent.kt:61)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$projectOpened$1.invokeSuspend(LanguageServerProjectComponent.kt:130)
	at (Coroutine boundary.()
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$projectOpened$1.invokeSuspend(LanguageServerProjectComponent.kt:130)
Caused by: com.intellij.util.IncorrectOperationException: Sorry but parent: Project (name=1.14 ***  is being disposed so the child: false will never be disposed
	at com.intellij.openapi.util.ObjectTree.register(ObjectTree.java:65)
	at com.intellij.openapi.util.Disposer.register(Disposer.java:78)
	at com.intellij.util.containers.DisposableWrapperList.createDisposableWrapper(DisposableWrapperList.java:255)
	at com.intellij.util.containers.DisposableWrapperList.add(DisposableWrapperList.java:62)
	at com.intellij.util.EventDispatcher.addListener(EventDispatcher.java:155)
	at com.intellij.openapi.editor.impl.event.EditorEventMulticasterImpl.addDocumentListener(EditorEventMulticasterImpl.java:120)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$setupServer$2.invokeSuspend(LanguageServerProjectComponent.kt:109)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$setupServer$2.invoke(LanguageServerProjectComponent.kt)
	at kotlinx.coroutines.intrinsics.UndispatchedKt.startUndispatchedOrReturn(Undispatched.kt:91)
	at kotlinx.coroutines.BuildersKt__Builders_commonKt.withContext(Builders.common.kt:156)
	at kotlinx.coroutines.BuildersKt.withContext(Unknown Source)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent.setupServer(LanguageServerProjectComponent.kt:61)
	at com.google.cloud.tools.intellij.languageclient.LanguageServerProjectComponent$projectOpened$1.invokeSuspend(LanguageServerProjectComponent.kt:130)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:740)
```

app.name.version
IntelliJ IDEA
app.version.major
2019
app.version.minor
3.1
java.version
1.8.0_202
java.vm.vendor
Oracle Corporation
os.name
Windows 10
plugin.version
20.1.1

**Comment by peterlin741** [comment]
`Issue 2694 - LanguageServerProjectComponent.setupServer uses disposed component`
`Fixes #2694`

The error seems to be coming from line 109 `eventMulticaster.addDocumentListener(editorWatcher, project)`. 

My initial thoughts are: 
- `LanguageServerProjectComponent::projectClosed`should remove listeners of a closed project (document listeners and editor listers) created in `setupServer`
- It's possible that reopening a closed project (closed projects are disposed?) reopens the editors (children of the parent project), which are linked to preexisting listeners

I found a link of another user's steps to roughly reproduce the same error, but I haven't had any luck:

https://youtrack.jetbrains.com/issue/MPS-25264
```
-  I closed the project I was working (it's under git, has commits, branches)
-  Opened attributes sample project from Welcome Screen
-  Main menu | VCS | Import into Version Control | Create git repository
-  Main menu | File | Open Recent | take first project in the list -> the project is opened, exception occurred
-  Blame dialog says the problem is in Git Integration plugin
```

Other users have ran into similar issues, but they appeared to be fixed in the IntelliJ source code: https://github.com/jk1/youtrack-idea-plugin/issues/67

**Comment by matthewmichihara** [comment]
I just noticed that the error is that a dispose is in progress (`DISPOSE_IN_PROGRESS`), and we're checking `project.isDisposed` which doesn't cover that case. I also just noticed that there is a `project.disposedOrDisposeInProgress` method that sounds useful? It looks like it is an experimental UI but it looks useful. If we can't reproduce this error though, it is hard to really verify.

**Comment by peterlin741** [comment]
Currently not able to reproduce this error

**Comment by etanshaul** [comment]
@peterlin741 closing this one for now as "not reproducible" Let's keep an eye out for these, and reopen if needed

**Comment by etanshaul** [comment]
seems we are still getting some of these. reopening for investigation

**Comment by ivanporty** [comment]
New reports have been received.

**Comment by jmah-cc** [comment]
@peterlin741 Should this issue be closed based on your linked fixed above?

**Comment by peterlin741** [comment]
Let's keep this open (the above was a similar, but different issue)

**Comment by peterlin741** [comment]
Fix in `20.8.1`, Can reopen if we see this again
